From 81650ae02ab973c08024b6ebf6811256a8e51e03 Mon Sep 17 00:00:00 2001
From: Fuad Tabba <tabba@google.com>
Date: Mon, 13 Apr 2026 14:13:31 +0000
Subject: [PATCH 094/223] ANDROID: KVM: arm64: Fix TRFCR_EL1 handling in
 __debug_save_trace

When the TRBE patch was backported, the TRFCR_EL1 zeroing was
incorrectly placed inside the TRBLIMITR_EL1.E check. This meant that
if TRBE was disabled, trace filtering was not disabled on guest entry,
and worse, an uninitialized/stale 0 value was unconditionally written
to TRFCR_EL1 on guest exit, breaking non-TRBE trace sinks.

This fixup aligns the logic with upstream V4, which unconditionally
reads and zeroes TRFCR_EL1 before conditionally draining TRBE.

Bug: 473214797
Fixes: dfb02af80897 ("BACKPORT: FROMLIST: KVM: arm64: Disable TRBE Trace Buffer Unit when running in guest context")
Link: https://lore.kernel.org/all/20260327130047.21065-2-will@kernel.org/
Change-Id: I8ed4558f099adddfc6f35feb5033f412dfb296c9
Signed-off-by: Fuad Tabba <tabba@google.com>
(cherry picked from commit 78864d9d283ee094c636f0226f109c61e382ec0f)
Bug: 504744760
---
 arch/arm64/kvm/hyp/nvhe/debug-sr.c | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/arch/arm64/kvm/hyp/nvhe/debug-sr.c b/arch/arm64/kvm/hyp/nvhe/debug-sr.c
index 99479b11857a..f83d852e6103 100644
--- a/arch/arm64/kvm/hyp/nvhe/debug-sr.c
+++ b/arch/arm64/kvm/hyp/nvhe/debug-sr.c
@@ -57,18 +57,17 @@ static void __debug_restore_spe(u64 pmscr_el1, u64 pmblimitr_el1)
 
 static void __debug_save_trace(u64 *trfcr_el1, u64 *trblimitr_el1)
 {
-	*trblimitr_el1 = read_sysreg_s(SYS_TRBLIMITR_EL1);
+	/*
+	 * Prohibit trace generation while we are in guest.
+	 * Since access to TRFCR_EL1 is trapped, the guest can't
+	 * modify the filtering set by the host.
+	 */
+	*trfcr_el1 = read_sysreg_el1(SYS_TRFCR);
+	write_sysreg_el1(0, SYS_TRFCR);
 
 	/* Check if the TRBE is enabled */
+	*trblimitr_el1 = read_sysreg_s(SYS_TRBLIMITR_EL1);
 	if (*trblimitr_el1 & TRBLIMITR_EL1_E) {
-		/*
-		 * Prohibit trace generation while we are in guest.
-		 * Since access to TRFCR_EL1 is trapped, the guest can't
-		 * modify the filtering set by the host.
-		 */
-		*trfcr_el1 = read_sysreg_el1(SYS_TRFCR);
-		write_sysreg_el1(0, SYS_TRFCR);
-
 		/*
 		 * The host has enabled the Trace Buffer Unit so we have
 		 * to beat the CPU with a stick until it stops accessing
-- 
2.50.1

